diff --git a/tensorflow/lite/micro/examples/hello_world/hello_world_test.cc b/tensorflow/lite/micro/examples/hello_world/hello_world_test.cc
index 7ced432..a8e45a3 100644
--- a/tensorflow/lite/micro/examples/hello_world/hello_world_test.cc
+++ b/tensorflow/lite/micro/examples/hello_world/hello_world_test.cc
@@ -13,6 +13,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "tensorflow/lite/micro/micro_optional_debug_tools.h"
 #include "tensorflow/lite/micro/all_ops_resolver.h"
 #include "tensorflow/lite/micro/examples/hello_world/model.h"
 #include "tensorflow/lite/micro/micro_error_reporter.h"
@@ -38,6 +39,8 @@ TF_LITE_MICRO_TEST(LoadModelAndPerformInference) {
                          model->version(), TFLITE_SCHEMA_VERSION);
   }
 
+  tflite::PrintModelData(model, &micro_error_reporter);
+
   // This pulls in all the operation implementations we need
   tflite::AllOpsResolver resolver;
 
@@ -57,6 +60,8 @@ TF_LITE_MICRO_TEST(LoadModelAndPerformInference) {
   // Allocate memory from the tensor_arena for the model's tensors
   TF_LITE_MICRO_EXPECT_EQ(interpreter.AllocateTensors(), kTfLiteOk);
 
+  tflite::PrintInterpreterState(&interpreter);
+
   // Alert for substantial increase in arena size usage.
   TF_LITE_MICRO_EXPECT_LE(interpreter.arena_used_bytes(),
                           model_arena_size + 100);
