diff --git a/tensorflow/lite/micro/testing/micro_test.h b/tensorflow/lite/micro/testing/micro_test.h
index d74d8f4..da5fcb0 100644
--- a/tensorflow/lite/micro/testing/micro_test.h
+++ b/tensorflow/lite/micro/testing/micro_test.h
@@ -65,6 +65,8 @@ extern bool did_test_fail;
 extern tflite::ErrorReporter* reporter;
 }  // namespace micro_test
 
+extern int test_main(int argc, char **argv);
+
 #define TF_LITE_MICRO_TESTS_BEGIN              \
   namespace micro_test {                       \
   int tests_passed;                            \
@@ -74,7 +76,7 @@ extern tflite::ErrorReporter* reporter;
   tflite::ErrorReporter* reporter;             \
   }                                            \
                                                \
-  int main(int argc, char** argv) {            \
+  int test_main(int argc, char** argv) {       \
     micro_test::tests_passed = 0;              \
     micro_test::tests_failed = 0;              \
     tflite::MicroErrorReporter error_reporter; \
