diff --git a/tensorflow/lite/kernels/internal/quantization_util.h b/tensorflow/lite/kernels/internal/quantization_util.h
index 0ee914b..52f3b14 100644
--- a/tensorflow/lite/kernels/internal/quantization_util.h
+++ b/tensorflow/lite/kernels/internal/quantization_util.h
@@ -133,7 +133,8 @@ IntOut SafeCast(FloatIn x) {
   static_assert(std::numeric_limits<IntOut>::radix == 2, "IntOut is base 2");
 
   // Special case NaN, for which the logic below doesn't work.
-  if (std::isnan(x)) {
+  // if (std::isnan(x)) {
+  if (isnan(x)) {
     return 0;
   }
 
@@ -143,7 +144,8 @@ IntOut SafeCast(FloatIn x) {
   }
 
   // Handle infinities.
-  if (std::isinf(x)) {
+  // if (std::isinf(x)) {
+  if (isinf(x)) {
     return x < 0 ? std::numeric_limits<IntOut>::min()
                  : std::numeric_limits<IntOut>::max();
   }
